lib/utils/compose_ts.ts

```diff
@@ -44,6 +45,73 @@ import type { DeviceInfo } from './device/api';
 import { getBalenaSdk, getChalk, stripIndent } from './lazy';
 import Logger = require('./logger');
 
+/**
+ * Return a set of build arguments passed in as flag options.
+ *
+ * @param buildArgs List of raw build argument strings
+ * @param composition Optional previously parsed composition object
+ */
+async function buildArgumentsFromInput(
```

Comment:
Throwing an error if the specified service does not exist seems like a sensible approach with the added benefit of possibly catching environment variable value confusion before it happens. In regards to colons in variable names, from what I understand it is considered an illegal character for both [Unix](https://stackoverflow.com/questions/2821043/allowed-characters-in-linux-environment-variable-names#:~:text=So%20names%20may%20contain%20any,not%20begin%20with%20a%20digit.) and [Windows](https://stackoverflow.com/questions/20635695/what-are-valid-characters-for-windows-environment-variable-names-and-values/20635858) environments. Therefore, I'm not sure we should worry about supporting such bad practices in balenaCLI.

And introducing this new behavior initially for `--buildArgs` alone, and then bringing it in to `--env` as a breaking change sounds reasonable. Unless you have any other concerns, I'll go ahead and update this function to throw errors if the specified service isn't defined in the project's composition.

Reply:
> In regards to colons in variable names, from what I understand it is considered an illegal character for both Unix and Windows environments

Hmm, the problem is that we got bitten by this before. I believe that bash and PowerShell don't allow the dot character in environment variable names, and yet a user needed to set an env var named `INTEGRATION.MQTT.AUTH.GENERIC.SERVER` in a docker-compose file and the supervisor code had to be fixed to allow it: https://github.com/balena-io/balena-supervisor/issues/1104

PowerShell on Windows seems happy to set env vars with a colon in the name, e.g. `MY:VAR`:
```
> $Env:MY:VAR=2

> get-item Env:*
...
MY:VAR        2
```

While `cmd.exe` on Windows is happy with colons and dots:

```
> set MY.VAR=2
> set MY:VAR=2
> set
...
MY.VAR=2
MY:VAR=2
```

And even if most _shells_ (rather than OS kernels) didn't support setting env vars with the dot or colon characters in the name, this doesn't prevent applications (written in most programming languages) setting and using such variables, as the OS kernel (or user space C libraries?) allow them. Node.js is quite happy with dots and colons:

```
> process.env['MY:VAR'] = 2
2
> process.env['INTEGRATION.MQTT.AUTH.GENERIC.SERVER'] = 'tcp://mosquitto:1883'
'tcp://mosquitto:1883'
> process.env
{
  ...
  'INTEGRATION.MQTT.AUTH.GENERIC.SERVER': 'tcp://mosquitto:1883',
  'MY:VAR': '2'
}
```

A Node.js application could set an env var containing colons and dots in the name, and directly pass these variables over to another application executed as a child process (the other app having been written in any other programming language), without getting shells involved.

So I am not sure about simply ignoring the possibility of the colon in environment variable names. It's tempting to say _"bash doesn't support it, we'll be fine",_ but then balena-io/balena-supervisor/issues/1104 is an actual issue that a balena user brought to support, not a hypothetical issue.

By the way, I've remembered about an open CLI issue regarding the `--env` parsing of the '=' character, which we could try to avoid or fix when parsing `--buildArg`: https://github.com/balena-io/balena-cli/issues/1705

Reply:
Oh wait. Build variables / args are not environment variables. `--env` needs to deal with environment variables, but `--buildArgs` only needs to accept whatever Docker / balenaEngine accepts:
* As the `buildargs` JSON object of the Docker API `ImageBuild` endpoint:  
https://docs.docker.com/engine/api/v1.41/#operation/ImageBuild
* As the ARG instruction in a Dockerfile:  
https://docs.docker.com/engine/reference/builder/#arg
* As the `--build-arg` flag of the docker CLI `docker build` command:  
https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg

So if we could determine that the colon character is not acceptable for some of the above, it would have more weight for `--buildArgs` than what an environment variable name may be!

Reply:
Tried adding in a colon in the name of a build arg and it doesn't look like it's properly supported? Still looking for something in the documentation to back this up:
```
$ docker build --build-arg MY_VAR=my_value --build-arg NPM:TOKEN=1234 -f Dockerfile .
...
Step 4/17 : ARG MY_VAR
 ---> Running in c7cc8ff2fadb
Removing intermediate container c7cc8ff2fadb
 ---> 211b07ad5d78
Step 5/17 : ARG NPM:TOKEN
 ---> Running in 7a414dae3c98
Removing intermediate container 7a414dae3c98
 ---> fac082db7be5
Step 6/17 : RUN echo "MY_VAR: $MY_VAR"
 ---> Running in 1da221eb092f
MY_VAR: my_value
Removing intermediate container 1da221eb092f
 ---> d2d072226751
Step 7/17 : RUN echo "NPM_TOKEN: $NPM:TOKEN"
 ---> Running in ee6819e2dd11
NPM_TOKEN: :TOKEN
...
```

```
$ docker build --build-arg "MY_VAR=my_value" --build-arg "NPM:TOKEN=1234" -f Dockerfile .
...
Step 4/17 : ARG MY_VAR
 ---> Running in b86167282979
Removing intermediate container b86167282979
 ---> b7bb783ebb0d
Step 5/17 : ARG NPM:TOKEN
 ---> Running in 43eb439b05f7
Removing intermediate container 43eb439b05f7
 ---> 93031a4872c1
Step 6/17 : RUN echo "MY_VAR: $MY_VAR"
 ---> Running in 0fefe6e2bbd2
MY_VAR: my_value
Removing intermediate container 0fefe6e2bbd2
 ---> d62579067df0
Step 7/17 : RUN echo "NPM_TOKEN: $NPM:TOKEN"
 ---> Running in 3e0de989512e
NPM_TOKEN: :TOKEN
...
```

Reply:
Also updated `buildArgumentsFromInput` to throw an error if an unknown service name was specified in a build argument instead of assigning it as a global build argument.

Reply:
Neither of the following would work:
```
RUN echo "NPM_TOKEN: $NPM:TOKEN"
RUN echo "NPM_TOKEN: ${NPM:TOKEN}"
```
because they use the "shell form" of RUN, where the line is executed through `/bin/sh -c` on Linux and `cmd /S /C` on Windows: https://docs.docker.com/engine/reference/builder/#run

And `sh` or `bash` don't like variables with a colon in the name. However, if use the "exec form" of RUN and, for example, execute `node` instead of `sh`, then `node` is OK with a variable with a colon in the name:

`Dockerfile`
```
FROM balenalib/intel-nuc-node
ARG MY_VAR
ARG NPM:TOKEN
RUN ["/usr/local/bin/node", "-e", "console.log(process.env)"]
RUN ["/usr/local/bin/node", "-e", "console.log(`NPM:TOKEN: ${process.env['NPM:TOKEN']}`)"]
CMD /bin/sh
```
```
$ docker build --build-arg 'NPM:TOKEN=1234' --no-cache .
...
Step 4/6 : RUN ["/usr/local/bin/node", "-e", "console.log(process.env)"]
 ---> Running in 87c47b8e3c01
{
  PATH: '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
  HOSTNAME: '2c99176dfd86',
  LC_ALL: 'C.UTF-8',
  DEBIAN_FRONTEND: 'noninteractive',
  UDEV: 'off',
  NODE_VERSION: '15.7.0',
  YARN_VERSION: '1.22.4',
  'NPM:TOKEN': '1234',
  HOME: '/root'
}

Step 5/6 : RUN ["/usr/local/bin/node", "-e", "console.log(`NPM:TOKEN: ${process.env['NPM:TOKEN']}`)"]

'NPM:TOKEN': 1234
```

Btw, while running this experiment, I've also found this interesting bit of Dockerfile documentation:
> https://docs.docker.com/engine/reference/builder/#using-arg-variables
> _You can use an ARG or an ENV instruction to specify variables that are available to the RUN instruction. Environment variables defined using the ENV instruction always override an ARG instruction of the same name._

So I think that this experiment has shown us a couple of things:
* `docker build --build-arg`, Dockerfile ARG, the Docker API and Docker Engine are happy with build ARG variables with a colon character in the name.
* My previous statement that _"Build variables / args are not environment variables"_ is actually not so true! Build ARG variables are actually provided as environment variables to processes spawned by RUN instructions in exec form, and ENV instructions simply override ARG instructions.

Reply:
Also:
* https://forums.balena.io/t/restrictions-for-keys-in-device-device-service-environment-variables/220331  
_"I want to use an application that uses environment variables in a similar form like this: foo:bar:key=somevalue"_